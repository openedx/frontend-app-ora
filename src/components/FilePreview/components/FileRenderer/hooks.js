import { useState } from 'react';
import { errorStatuses, errorMessages, renderers } from '../constants';
import { getFileType } from '../utils';
import messages from './messages';

/**
 * component hooks
 */
export const useRenderData = ({
  file,
  formatMessage,
}) => {
  const [errorStatus, setErrorStatus] = useState(null);
  const [isLoading, setIsLoading] = useState(true);

  const setState = (newState) => {
    setErrorStatus(newState.errorStatus);
    setIsLoading(newState.isLoading);
  };

  const stopLoading = (status = null) => setState({ isLoading: false, errorStatus: status });

  const errorMessage = (
    errorMessages[errorStatus] || errorMessages[errorStatuses.serverError]
  );
  const errorAction = {
    id: 'retry',
    onClick: () => setState({ errorStatus: null, isLoading: true }),
    message: messages.retryButton,
  };
  const error = {
    headerMessage: errorMessage,
    children: formatMessage(errorMessage),
    actions: [errorAction],
  };

  const Renderer = renderers[getFileType(file.fileName)];
  const rendererProps = {
    fileName: file.fileName,
    url: file.fileUrl,
    onError: stopLoading,
    onSuccess: () => stopLoading(),
  };

  return {
    errorStatus,
    isLoading,
    error,
    Renderer,
    rendererProps,
  };
};
