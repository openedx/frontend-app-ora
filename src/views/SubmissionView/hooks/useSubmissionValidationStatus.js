import { useCallback } from 'react';
import { StrictDict, useKeyedState } from '@edx/react-unit-test-utils';

import {
  useSubmissionConfig,
} from 'hooks/app';

export const stateKeys = StrictDict({
  promptStatuses: 'promptStatuses',
  submissionTriggered: 'submissionTriggered',
  fileUploadIsRequired: 'fileUploadIsRequired',
});

const useSubmissionStatuses = (textResponses, uploadedFiles) => {
  const submissionConfig = useSubmissionConfig();

  // holds the status of each text field
  const [promptStatuses, setPromptStatuses] = useKeyedState(
    stateKeys.promptStatuses,
    textResponses.reduce((obj, val, i) => ({ ...obj, [i]: true }), {}),
  );
  const [submissionTriggered, setSubmissionTriggered] = useKeyedState(stateKeys.submissionTriggered, false);
  const [fileUploadIsRequired, setFileUploadIsRequired] = useKeyedState(stateKeys.fileUploadIsRequired, false);

  // this function is to be called when response submit button is clicked
  const validateBeforeConfirmation = useCallback(() => {
    setSubmissionTriggered(true);
    const retrievePromptsSubmissionStatus = textResponses.reduce((obj, val, i) => ({ ...obj, [i]: val }), {});
    let containsInvalidPrompts = false;
    if (submissionConfig.textResponseConfig.required) {
      for (let i = 0; i < textResponses.length; i++) {
        retrievePromptsSubmissionStatus[i] = !!textResponses[i];
      }
      containsInvalidPrompts = Object.values(retrievePromptsSubmissionStatus).some((val) => !val);
    }
    const calcFileUploadIsRequired = submissionConfig.fileResponseConfig.required && !uploadedFiles.length;
    setFileUploadIsRequired(calcFileUploadIsRequired);
    if (containsInvalidPrompts || (submissionConfig.fileResponseConfig.required && !uploadedFiles.length)) {
      setPromptStatuses(retrievePromptsSubmissionStatus);
      return false;
    }

    return true;
  }, [
    textResponses,
    uploadedFiles,
    submissionConfig,
    setPromptStatuses,
    setSubmissionTriggered,
    setFileUploadIsRequired,
  ]);

  return {
    validateBeforeConfirmation,
    setSubmissionTriggered,
    submissionTriggered,
    setPromptStatuses,
    promptStatuses,
    fileUploadIsRequired,
  };
};

export default useSubmissionStatuses;
