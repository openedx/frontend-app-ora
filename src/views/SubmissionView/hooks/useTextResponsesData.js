import { useCallback, useEffect } from 'react';

import { StrictDict, useKeyedState } from '@edx/react-unit-test-utils';
import {
  useFinishLater,
  useHasSubmitted,
  useSaveDraftResponse,
  useTextResponses,
} from 'hooks/app';
import { MutationStatus } from 'constants/index';

export const stateKeys = StrictDict({
  textResponses: 'textResponses',
  isDirty: 'isDirty',
  hasSaved: 'hasSaved',
  lastChanged: 'lastChanged',
});

const useTextResponsesData = ({ setHasSavedDraft }) => {
  const textResponses = useTextResponses();
  const hasSubmitted = useHasSubmitted();
  const [hasSaved, setHasSaved] = useKeyedState(stateKeys.hasSaved, false);
  const [lastChanged, setLastChanged] = useKeyedState(stateKeys.lastChanged, null);
  const [isDirty, setIsDirty] = useKeyedState(stateKeys.isDirty, false);
  const [value, setValue] = useKeyedState(stateKeys.textResponses, textResponses);

  const saveResponseMutation = useSaveDraftResponse();
  const finishLaterMutation = useFinishLater();

  const saveResponse = useCallback(() => {
    setIsDirty(false);
    return saveResponseMutation.mutateAsync({ textResponses: value }).then(() => {
      setHasSavedDraft(true);
    });
  }, [
    saveResponseMutation,
    value,
    setIsDirty,
    setHasSavedDraft,
  ]);

  const finishLater = useCallback(() => {
    setIsDirty(false);
    return finishLaterMutation.mutateAsync({ textResponses: value });
  }, [setIsDirty, finishLaterMutation, value]);

  const onChange = useCallback((index) => (textResponse) => {
    setValue(oldResponses => {
      const out = [...oldResponses];
      out[index] = textResponse;
      return out;
    });
    setIsDirty(true);
    setLastChanged(Date.now());
  }, [
    setValue,
    setIsDirty,
    setLastChanged,
  ]);

  useEffect(() => {
    const interval = setInterval(() => {
      if (!lastChanged) {
        return;
      }
      const timeSinceChange = Date.now() - lastChanged;
      if (isDirty && timeSinceChange > 2000) {
        saveResponse();
        if (!hasSaved) {
          setHasSaved(true);
        }
      }
    }, 2000);
    return () => clearInterval(interval);
  }, [
    isDirty,
    hasSubmitted,
    hasSaved,
    setHasSaved,
    lastChanged,
    saveResponse,
  ]);

  return {
    textResponses: value,
    onUpdateTextResponse: onChange,
    isDraftSaved: saveResponseMutation.status === MutationStatus.success && !isDirty,
    saveResponse,
    saveResponseStatus: saveResponseMutation.status,
    finishLater,
    finishLaterStatus: finishLaterMutation.status,
  };
};

export default useTextResponsesData;
