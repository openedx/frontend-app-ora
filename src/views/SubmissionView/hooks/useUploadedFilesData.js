import React from 'react';

import { StrictDict, useKeyedState } from '@edx/react-unit-test-utils';

import {
  useResponseData,
  useUploadFiles,
  useDeleteFile,
} from 'hooks/app';

export const stateKeys = StrictDict({ uploadedFiles: 'uploadedFiles' });

const useUploadedFilesData = () => {
  const deleteFileMutation = useDeleteFile();
  const uploadFilesMutation = useUploadFiles();

  const response = useResponseData() || {};

  const [value, setValue] = useKeyedState(
    stateKeys.uploadedFiles,
    response?.uploadedFiles ? response.uploadedFiles : [],
  );

  React.useEffect(() => {
    setValue(response.uploadedFiles);
  }, [setValue, response.uploadedFiles]);

  const onFileUploaded = uploadFilesMutation.mutateAsync;
  const onDeletedFile = deleteFileMutation.mutateAsync;

  return {
    uploadedFiles: value,
    onFileUploaded,
    onDeletedFile,
  };
};

export default useUploadedFilesData;
