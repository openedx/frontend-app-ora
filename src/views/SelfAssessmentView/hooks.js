import { useORAConfigData, usePageData } from 'data/services/lms/hooks/selectors';

const useAssessmentViewHooks = () => {
  const pageData = usePageData();
  const oraConfigData = useORAConfigData();

  return {
    submission: pageData?.submission,
    oraConfigData,
  };
};

export default useAssessmentViewHooks;
