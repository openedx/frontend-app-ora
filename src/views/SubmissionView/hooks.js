import { useEffect, useReducer } from 'react';

import { useORAConfigData, usePageData } from 'data/services/lms/hooks/selectors';

import { submitResponse, saveResponse, uploadFiles } from 'data/services/lms/hooks/actions';
import { MutationStatus } from 'data/services/lms/constants';

const useSubmissionViewHooks = () => {
  const submitResponseMutation = submitResponse();
  const saveResponseMutation = saveResponse();
  const uploadFilesMutation = uploadFiles();
  const pageData = usePageData();
  const oraConfigData = useORAConfigData();

  const [submission, dispatchSubmission] = useReducer(
    (state, payload) => ({ ...state, isDirty: true, ...payload }),
    { ...pageData?.submission, isDirty: false },
  );

  useEffect(() => {
    // a workaround to update the submission state when the pageData changes
    if (pageData?.submission) {
      dispatchSubmission({ ...pageData.submission, isDirty: false });
    }
  }, [pageData?.submission]);

  const onTextResponseChange = (index) => (textResponse) => {
    dispatchSubmission({
      response: {
        ...submission.response,
        textResponses: [
          ...submission.response.textResponses.slice(0, index),
          textResponse,
          ...submission.response.textResponses.slice(index + 1),
        ],
      },
    });
  };

  const onFileUploaded = (args) => {
    const fileUploads = uploadFilesMutation.mutate(args);
    dispatchSubmission({ response: { ...submission.response, fileUploads } });
  };

  const submitResponseHandler = () => {
    dispatchSubmission({ isDirty: false });
    submitResponseMutation.mutate(submission);
  };

  const saveResponseHandler = () => {
    dispatchSubmission({ isDirty: false });
    saveResponseMutation.mutate(submission);
  };

  return {
    submitResponseHandler,
    submitResponseStatus: submitResponseMutation.status,
    saveResponseHandler,
    saveResponseStatus: saveResponseMutation.status,
    draftSaved: saveResponseMutation.status === MutationStatus.success && !submission.isDirty,
    pageData,
    oraConfigData,
    submission,
    dispatchSubmission,
    onTextResponseChange,
    onFileUploaded,
  };
};

export default useSubmissionViewHooks;
