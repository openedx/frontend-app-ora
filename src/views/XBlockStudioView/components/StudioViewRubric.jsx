import React from 'react';

import { useIntl } from '@edx/frontend-platform/i18n';
import { Collapsible } from '@edx/paragon';
import { useRubricConfig } from 'hooks/app';

import { useXBlockStudioViewContext } from './XBlockStudioViewProvider';
import messages from './messages';

const StudioViewRubric = () => {
  const { formatMessage } = useIntl();
  const { criteria } = useRubricConfig();
  const { rubricIsOpen, toggleRubric } = useXBlockStudioViewContext();

  return (
    <Collapsible
      title={<h3 className="py-3">{formatMessage(messages.rubricHeader)}</h3>}
      open={rubricIsOpen}
      onToggle={toggleRubric}
    >
      {criteria.map((criterion) => (
        <div key={criterion.name} className="py-3">
          <p>
            <strong>{formatMessage(messages.criteriaNameLabel)}</strong>
            {criterion.name}
          </p>
          <p>
            <strong>{formatMessage(messages.criteriaDescriptionLabel)}</strong>
            {criterion.description}
          </p>
          <p>
            <strong>{formatMessage(messages.criteriaOptionsLabel)}</strong>
          </p>
          <ul>
            {criterion.options.map((option) => (
              <li key={option.name}>
                <strong>
                  {option.label}: {option.points} {formatMessage(messages.pointsLabel)}
                </strong>
                <p>{option.description}</p>
              </li>
            ))}
          </ul>
        </div>
      ))}
    </Collapsible>
  );
};

export default StudioViewRubric;
