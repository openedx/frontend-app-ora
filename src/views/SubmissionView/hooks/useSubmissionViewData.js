import React from 'react';

import { StrictDict, useKeyedState } from '@edx/react-unit-test-utils';

import {
  useGlobalState,
  useRubricConfig,
  useSubmitResponse,
  useSetHasSubmitted,
  useHasSubmitted,
  useRefreshPageData,
  useSetResponse,
  useResponse,
} from 'hooks/app';
import {
  useRefreshUpstream,
} from 'hooks/modal';
import { stepStates, stepNames } from 'constants/index';

import useTextResponsesData from './useTextResponsesData';
import useUploadedFilesData from './useUploadedFilesData';

const stateKeys = StrictDict({
  hasSavedDraft: 'hasSavedDraft',
});

const useSubmissionViewData = () => {
  const [hasSavedDraft, setHasSavedDraft] = useKeyedState(stateKeys.hasSavedDraft, false);
  const hasSubmitted = useHasSubmitted();
  const setHasSubmitted = useSetHasSubmitted();
  const submitResponseMutation = useSubmitResponse();
  const rubricConfig = useRubricConfig();
  const globalState = useGlobalState({ step: stepNames.submission });
  const refreshPageData = useRefreshPageData();
  const setResponse = useSetResponse();
  const response = useResponse();
  const refreshUpstream = useRefreshUpstream();
  const stepState = hasSubmitted ? stepStates.submitted : globalState.stepState;

  const {
    textResponses,
    onUpdateTextResponse,
    saveResponse,
    saveResponseStatus,
    finishLater,
    finishLaterStatus,
  } = useTextResponsesData({ setHasSavedDraft });

  const {
    uploadedFiles,
    onFileUploaded,
    onDeletedFile,
  } = useUploadedFilesData();

  const submitResponseHandler = React.useCallback(() => {
    submitResponseMutation.mutateAsync({
      textResponses,
      uploadedFiles,
    }).then(() => {
      setResponse({ textResponses, uploadedFiles });
      setHasSubmitted(true);
      refreshPageData();
      refreshUpstream();
    });
  }, [ // eslint-disable-line react-hooks/exhaustive-deps
    setHasSubmitted,
    submitResponseMutation,
    textResponses,
    uploadedFiles,
  ]);

  return {
    actionOptions: {
      finishLater,
      finishLaterStatus,
      saveResponse,
      saveResponseStatus,
      submit: submitResponseHandler,
      submitStatus: submitResponseMutation.status,
      hasSubmitted,
    },
    response: hasSubmitted
      ? response
      : { textResponses, uploadedFiles },
    hasSubmitted,
    onUpdateTextResponse,
    isDraftSaved: hasSavedDraft,
    onDeletedFile,
    onFileUploaded,
    showRubric: rubricConfig.showDuringResponse,
    isReadOnly: stepState === stepStates.done || hasSubmitted,
  };
};

export default useSubmissionViewData;
