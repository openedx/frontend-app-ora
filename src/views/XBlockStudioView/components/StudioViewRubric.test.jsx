import { shallow } from '@edx/react-unit-test-utils';

import { useRubricConfig } from 'hooks/app';

import StudioViewRubric from './StudioViewRubric';

jest.mock('hooks/app', () => ({
  useRubricConfig: jest.fn(),
}));
jest.mock('./XBlockStudioViewProvider', () => ({
  useXBlockStudioViewContext: () => ({
    rubricIsOpen: true,
    toggleRubric: jest.fn().mockName('toggleRubric'),
  }),
}));

describe('<StudioViewRubric />', () => {
  it('render with criteria and options', () => {
    useRubricConfig.mockReturnValue({
      criteria: [
        {
          name: 'criterion1',
          description: 'description1',
          options: [
            {
              name: 'option1',
              label: 'label1',
              points: 1,
              description: 'description1',
            },
            {
              name: 'option2',
              label: 'label2',
              points: 2,
              description: 'description2',
            },
          ],
        },
        {
          name: 'criterion2',
          description: 'description2',
          options: [
            {
              name: 'option2',
              label: 'label2',
              points: 2,
              description: 'description2',
            },
          ],
        },
      ],
    });

    const wrapper = shallow(<StudioViewRubric />);
    expect(wrapper.snapshot).toMatchSnapshot();

    expect(wrapper.instance.findByTestId('criteria-test-id').length).toBe(2);
  });

  it('render without criteria', () => {
    useRubricConfig.mockReturnValue({ criteria: [] });

    const wrapper = shallow(<StudioViewRubric />);
    expect(wrapper.snapshot).toMatchSnapshot();

    expect(wrapper.instance.findByTestId('criteria-test-id').length).toBe(0);
  });
});
