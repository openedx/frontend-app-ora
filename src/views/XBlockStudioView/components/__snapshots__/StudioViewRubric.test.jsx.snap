// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StudioViewRubric /> render with criteria and options 1`] = `
<Collapsible
  onToggle={[MockFunction toggleRubric]}
  open={true}
  title={
    <h3
      className="py-3"
    >
      Rubric
    </h3>
  }
>
  <div
    className="py-3"
    data-testid="criteria-test-id"
    key="criterion1"
  >
    <p>
      <strong>
        Criteria name: 
      </strong>
      criterion1
    </p>
    <p>
      <strong>
        Criteria description: 
      </strong>
      description1
    </p>
    <p>
      <strong>
        Criteria options: 
      </strong>
    </p>
    <ul>
      <li
        key="option1"
      >
        <strong>
          label1
          : 
          1
           
          points
        </strong>
        <p>
          description1
        </p>
      </li>
      <li
        key="option2"
      >
        <strong>
          label2
          : 
          2
           
          points
        </strong>
        <p>
          description2
        </p>
      </li>
    </ul>
  </div>
  <div
    className="py-3"
    data-testid="criteria-test-id"
    key="criterion2"
  >
    <p>
      <strong>
        Criteria name: 
      </strong>
      criterion2
    </p>
    <p>
      <strong>
        Criteria description: 
      </strong>
      description2
    </p>
    <p>
      <strong>
        Criteria options: 
      </strong>
    </p>
    <ul>
      <li
        key="option2"
      >
        <strong>
          label2
          : 
          2
           
          points
        </strong>
        <p>
          description2
        </p>
      </li>
    </ul>
  </div>
</Collapsible>
`;

exports[`<StudioViewRubric /> render without criteria 1`] = `
<Collapsible
  onToggle={[MockFunction toggleRubric]}
  open={true}
  title={
    <h3
      className="py-3"
    >
      Rubric
    </h3>
  }
/>
`;
